modules/dnn/src/graph_buffer_allocator.cpp

```diff
@@ -0,0 +1,331 @@
+                auto subgraphs = layer->subgraphs();
+                CV_Assert(subgraphs && subgraphs->size() == 1);
+                const Ptr<Graph>& body = subgraphs->at(0);
+                Arg trip_count = inputs[0];
+                const std::vector<Arg>& body_inputs = body->inputs();
+                const std::vector<Arg>& body_outputs = body->outputs();
+                size_t body_ninputs = body_inputs.size();
+                size_t body_noutputs = body_outputs.size();
+                int n_state_vars = (int)(ninputs - 2);
+                int n_accums = (int)(body_noutputs - n_state_vars - 1);
+                CV_Assert(body_ninputs == ninputs);
+                CV_Assert(body_noutputs == noutputs+1);
+                CV_Assert(n_state_vars >= 0 && n_accums >= 0);
+                Arg inp0 = inputs[0];
+                if (inp0.idx > 0 && usecounts[inp0.idx] > 0) {
+                    CV_Assert(!net->isConstArg(inp0));
+                    if (!net->isConstArg(trip_count))
+                        shareBuffer(trip_count, inputs[0]);
+                    else
+                        bufidxs.at(inputs[0].idx) = getFreeBuffer();
+                }
+
+                for (int i = -1; i < n_state_vars; i++) {
+                    Arg inparg = body_inputs[i+2];
+                    Arg outarg = body_outputs[i+1];
+                    Arg v_inp = inputs[i+2];
+                    Arg v_out = i >= 0 ? outputs[i] : Arg();
+                    if (inparg.idx > 0 && usecounts[inparg.idx] > 0) {
+                        CV_Assert(!net->isConstArg(inparg));
+                        if (!net->isConstArg(v_inp))
+                            shareBuffer(v_inp, inparg);
+                        else
+                            bufidxs[inparg.idx] = getFreeBuffer();
+                    }
+                    if (!net->isConstArg(v_out)) {
+                        if (!net->isConstArg(outarg) && usecounts[outarg.idx] == 1)
+                            shareBuffer(v_out, outarg);
+                    }
+                }
+
+                assign(body);
+                for (auto body_out: body_outputs)
+                    releaseBuffer(bufidxs.at(body_out.idx));
+            }
+
+            for (auto outarg: outputs) {
+                if (usecounts[outarg.idx] == 0)
+                    releaseBuffer(bufidxs.at(outarg.idx));
+            }
+            // let's release inputs in the reverse order to keep the buffer allocation consistent across the network
+            for (size_t i = 0; i < ninputs; i++) {
+                Arg inparg = inputs[ninputs-i-1];
+                int bufidx = bufidxs[inparg.idx];
+                if (bufidx >= 0) {
+                    if (--usecounts.at(inparg.idx) == 0)
+                        releaseBuffer(bufidx);
+                }
+            }
+        }
+    }
+};
+
+void Net::Impl::assignBuffers()
+{
+    BufferAllocator buf_allocator(net);
+    buf_allocator.assign();
```

Comment:
How to disable that? E.g for debug pursposes.

Reply:
buffer allocation (or rather a buffer assignment) is crucial and necessary preprocessing stage in the new engine. Without it you cannot call forward. Some other stages, such as fusion, could be potentially disabled; this is TODO.